source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -797,7 +797,7 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
 
   if (event == Network::ConnectionEvent::RemoteClose ||
       event == Network::ConnectionEvent::LocalClose) {
-    upstream_.reset();
+    read_callbacks_->connection().dispatcher().deferredDelete(std::move(upstream_));
```

Comment:
This line is highly suspected. We need to carefully think `deferedDelete()`.

Reply:
Any change to lifetime should have been runtime guarded, and any change to the data plane ought to be reviewed by a senior maintainer :-(